.github/workflows/build.yml

```diff
@@ -0,0 +1,195 @@
+  DEB_SCAN_BUILD_DEPS: |
+    clang
+    clang-tools
+  # Useful URL: https://git.archlinux.org/svntogit/community.git/tree/mate-panel
+  # + mate-desktop dependencies
+  ARCH_BUILD_DEPS: |
+    ccache
+    autoconf-archive
+    clang
+    gcc
+    git
+    gobject-introspection
+    gtk-layer-shell
+    itstool
+    libcanberra
+    libmateweather
+    libsm
+    libwnck3
+    make
+    mate-common
+    mate-desktop
+    mate-menus
+    meson
+    which
+    yelp-tools
+    iso-codes
+    gobject-introspection
+
+jobs:
+  build:
+    name: Build on ${{matrix.container}} with in-process=${{matrix.in-process}}
+    runs-on: ubuntu-latest
+    container: ${{matrix.container}}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        in-process: [all, none]
+        container: ['debian:testing', 'ubuntu:rolling', 'archlinux:latest']
+
+    env:
+      CC: ccache gcc
+      CXX: ccache g++
+      BASE_INSTALL_PATH: ${{github.workspace}}/_install
+
+    steps:
+      - name: Setup environment
+        run: |
+          echo "PATH=${BASE_INSTALL_PATH}/bin:${PATH}" >> "$GITHUB_ENV"
+          echo "PKG_CONFIG_PATH=${BASE_INSTALL_PATH}/lib/pkgconfig:${PKG_CONFIG_PATH}" >> "$GITHUB_ENV"
+
+      - name: Show environment
+        run: env | sort
+        if: ${{ env.DEBUG == '1' }}
+
+      # For Debian and Ubuntu (apt-based with reasonably compatible packages)
+      - name: Install dependencies
+        if: ${{ startsWith(matrix.container, 'debian:') || startsWith(matrix.container, 'ubuntu:') }}
+        run: |
+          apt-get update -qq
+          apt-get install --assume-yes --no-install-recommends \
+            ${DEB_BUILD_DEPS} ${DEB_LIBRARY_DEPS}
+
+      # For ArchLinux
+      - name: Install dependencies
+        if: ${{ startsWith(matrix.container, 'archlinux:') }}
+        # don't upgrade, although told otherwise (see link below), because
+        # apparently in the container it doesn't quit work...
+        # https://wiki.archlinux.org/title/System_maintenance#Partial_upgrades_are_unsupported
+        run: |
+          pacman --noconfirm -Sy
+          pacman --noconfirm -S ${ARCH_BUILD_DEPS}
+
+      - uses: actions/checkout@v3
+        with:
+          path: mate-panel
+          submodules: true
+
+      # TODO: cache mate-desktop build/install?
+      - name: Checkout mate-desktop
+        uses: actions/checkout@v3
+        with:
+          repository: mate-desktop/mate-desktop
+          ref: v1.27.1
+          path: mate-desktop
+          submodules: true
+
+      - name: ccache
+        uses: hendrikmuhs/ccache-action@v1.2
+
+      - name: Configure mate-desktop
+        run: |
+          cd mate-desktop
+          NOCONFIGURE=1 ./autogen.sh
+          { ./configure --prefix="$BASE_INSTALL_PATH" || { cat config.log; exit 1; } ; }
+
+      - name: Build mate-desktop
+        run: make -C mate-desktop -j ${{ env.JOBS }}
+
+      - name: Install mate-desktop
+        run: make -C mate-desktop -j ${{ env.JOBS }} install
```

Comment:
Ideally we'd cache all this, because it usually won't matter (well, until the distro updates in an incompatible manner maybe…)

Reply:
Building a dep installer package might be a solution.
In gitlab-ci there was a caching option iwhre you could cache a result until it was overwritten.
githup and gitlab CI is very similar . i will search for this
 .

Reply:
I found this cache for actions https://github.com/actions/cache
reading.......

Reply:
I think this can be used. Actions  `artifacts` or `cache`. see https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#comparing-artifacts-and-dependency-caching

Reply:
Yeah I know it's doable, I just didn't spend the time yet to understand how to use it here, and it looks OK but not entirely trivial

Reply:
I think we have to use the cache option, which cache a result from master (main) branch to all child branches in a directory. This result can be restored with a key.
But they explained it better than me with examples.

Reply:
Thanks, I finally managed to get some caching working with the documentation you linked.
I hope the cache key is precise enough not to cause issues later (esp. will it be OK with rolling release distros?), but we can always change the cache key or drop it altogether it it causes issues.

But caching did shave off a bit of build time :)

Reply:
> ............ I hope the cache key is precise enough not to cause issues later (esp. will it be OK with rolling release distros?), but we can always change the cache key or drop it altogether it it causes issues.
> 
> But caching did shave off a bit of build time :)

Amazing :)
Using a developer release as build dependency is really a special case.
When ubuntu:rolling use a new mate major release eg. 1.28 we need to update github-ci config, and split out the updated distros in a own config part.
Or we need a condition in configuration. Something like....
`If package in repos >= 1.28 than do not use cached package`
Not sure if this is possible......only an idea.
With travis-ci we had the same issue in past. CI is a static config which needs interaction from time to time.
Usually Archlinux is the first distro (we don't use fedora-rawhide) which ship new Mate releases or new packages releases which can cause surprises.